v4/wrapper.go

```diff
@@ -5,28 +5,44 @@ import (
 	"time"
 
 	"github.com/KolbyMcGarrah/ocredis"
+	"go.opencensus.io/trace"
 	pkgredis "gopkg.in/redis.v4"
 )
 
 // Wrap returns a wrapped redis client
-func Wrap(client *pkgredis.Client, instanceName string) *Wrapper {
+func Wrap(c *pkgredis.Client, options ...ocredis.TraceOption) *Wrapper {
+	o := ocredis.TraceOptions{}
+	for _, option := range options {
+		option(&o)
+	}
+	if o.InstanceName == "" {
+		o.InstanceName = ocredis.DefaultInstanceName
+	} else {
+		o.DefaultAttributes = append(o.DefaultAttributes, trace.StringAttribute("cache.instance", o.InstanceName))
+	}
 	return &Wrapper{
-		client:       client,
-		instanceName: instanceName,
+		client:  c,
+		options: o,
 	}
 }
 
 var _ ocredis.Client = &Wrapper{}
 
 // Wrapper wraps the redis package with an instance name to be used to collect metrics.
 type Wrapper struct {
-	client       *pkgredis.Client
-	instanceName string
+	client  *pkgredis.Client
+	options ocredis.TraceOptions
 }
 
 // Get integrates the redis get command with metrics
 func (w *Wrapper) Get(ctx context.Context, key string) (cmd ocredis.StringCmd) {
-	var recordCallFunc = ocredis.RecordCall(ctx, "go.redis.get", w.instanceName)
+	if ocredis.AllowTrace(ctx, w.options.Get, w.options.AllowRoot) {
+		span := ocredis.StartSpan(ctx, "Get", w.options)
+		if span != nil {
+			defer span.EndSpanWithErr(cmd.Err())
```

Comment:
This needs an update:
```suggestion
			defer func() { span.EndSpanWithErr(cmd.Err()) }()
```

There is a very sneaky caveat of `defer` which is:
> **The deferred call's arguments are evaluated immediately**, but the function call is not executed until the surrounding function returns.

This means that `cmd.Err()` is evaluated immediately which will always result in a call of `span.EndSpanWithErr(nil)` because at L42, `cmd` is the empty value of the `ocredis.StringCmd` struct.

Here is a (pretty confusing) example of this in the Go Playground. The point is that the first defer always prints `<nil>` instead of the error that is assigned to `foo`.
https://play.golang.org/p/YnC6KlvT-uy

Reply:
I think I fixed this by making an anonymous function for the defer which then calls the EndSpanWithErr inside of that. Pushing that up now.